src/graph_visualization.cpp

```diff
@@ -0,0 +1,449 @@
+#include <SDL2/SDL.h>
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+#define SCREEN_WIDTH 640
+#define SCREEN_HEIGHT 480
+#define DISTANCE 40
+
+struct tree_node {
+    using node_id = int;
+    vector<node_id> succ = {};
+    vector<node_id> pred = {};
+    node_id id = 0; // no
+    int deg_p = 0;
+    int deg_m = 0;
+    int layer = 0;
+    int number = 0;
+    int x = 0;
+    int y = 0;
+};
+
+struct Net {
+    using node_id = tree_node::node_id;
+    
+    const vector<node_id> &get_sources();
+    const vector<node_id> &get_sinks();
+    
+    const tree_node *get_node(node_id id);
+};
+
+vector< vector<int> > deleted_edges = {};
+
+vector< vector<int> > input_graph();
+vector<tree_node> id_succ_init( vector< vector<int> > input );
+vector<tree_node> pred_reinit( vector<tree_node> nodes );
+vector<tree_node> deg_reinit( vector<tree_node> nodes );
+void print_info( vector<tree_node> nodes );
+
+void DFS( int &exit, vector<tree_node> nodes, vector<bool> used_id, tree_node::node_id id, tree_node::node_id prev_id );
+void DFS_cycle_exists( vector<tree_node> nodes );
+
+vector<tree_node> remove_incident_arcs( vector<tree_node> nodes, tree_node::node_id id );
+
+// void topolog_sort( vector<tree_node> &nodes );
+void heapify( vector<tree_node> &nodes, int n, int i );
+void topolog_heapSort( vector<tree_node> &nodes );
+vector<tree_node> greedyFAS( vector<tree_node> nodes );
+
+vector<tree_node> ASAP( vector<tree_node> nodes );
+vector<tree_node> coord_init( vector<tree_node> nodes );
```

Comment:
Considering input networks are of non-trivial sizes the API forces excessive amount of copying. Rework the API to avoid copying of memory-allocating data structures, e.g. `std::vector`.